### ISSUE TITLE
[BUG] Cannot push artifacts over 4MB to Github Packages

### ISSUE THREAD

**Issue by brooksmtownsend** [krustlet/oci-distribution#35]
## Bug
Attempting to push Wasm or other OCI artifacts that are not containers to Github Packages seems to fail with an error when the artifact is over 4MB. We have no issues using this crate + `wash` to push to the localhost docker registry or AzureCR so this appears to be a Github CR issue. Files that are under 4MB do not have this issue. The specific error is:
```
An unexpected error occured: code=416 Range Not Satisfiable, message='{"errors":[{"code":"REQUESTED_RANGE_NOT_SATISFIABLE","message":"the request body is too large and exceeds the maximum permissible limit of 4.00MiB"}]}
'
```

## Detail 
I'm confused as to why this is an issue, as even the GHCR tutorial shows a container image that is 90+MB in size. I can only assume that this is due to the fact that we shove our actor and provider bytes into a single layer and that in chunks greater than 4MB.

I've attempted to engineer around this and split the artifact up into layers in https://github.com/wasmCloud/wash/pull/269, though I get a different error message (this is an example with two layers):
```
An unexpected error occured: code=400 Bad Request, message='{"errors":[{"code":"MANIFEST_BLOB_UNKNOWN","message":"blob unknown to registry","detail":"sha256:09adb5f67d938b7cd354bba834fc6df75f7325fdd67254aa14c4ac2390dc6727"},{"code":"MANIFEST_BLOB_UNKNOWN","message":"blob unknown to registry","detail":"sha256:0a708e592bb06968b84077a71296b6e1be19aad7d5556cd2b8d9fe1cd3ac6b7e"}]}
'
```

From the spec, I see that this can be occur if the manifest is not known to the registry. This was only an initial attempt, but in the interest of saving myself from missing a small detail in the OCI spec I wanted to post for some help. If this splitting into layers logic is something that would be useful upstream, once we figure this out I'd be happy to contribute that to this crate instead of in the wasmCloud shell.

### Additional context
Issue with wasmCloud specific detail: https://github.com/wasmCloud/wash/issues/26

**Comment by brooksmtownsend** [comment]
I am not _entirely_ sure what was wrong with my initial attempt, but with https://github.com/wasmCloud/wash/pull/269 I was able to resolve this on version `v0.9.0`. It works 🎉

**Comment by brooksmtownsend** [comment]
@flavio tagging you just because you reacted, if that helper logic to split an artifact into multiple layers with a certain size would be useful, I'd be happy to contribute that upstream

**Comment by flavio** [comment]
In theory this shouldn't be necessary. An OCI layer doesn't have size limits, hence the solution of splitting your binary blob into smaller layers is something you should not have to do.

I thought some of the fixes that went into `v0.9.0` addressed the issue...

I'm going to reopen the issue, to look more into that.

**Comment by brooksmtownsend** [comment]
It's entirely possible this is a specific, undocumented implementation detail with GitHub container registry and OCI artifacts

**Comment by flavio** [comment]
I see, just out of curiosity... have you tried to upload the same Wasm file using https://github.com/engineerd/wasm-to-oci ? Does it work using it?

I'm asking because wasm-to-oci leverage the same Go libraries used by docker, which proved to be more permissive with the OCI standard...

**Comment by brooksmtownsend** [comment]
I have not tried that, also the files that commonly exceed 4mb for us are our "provider archives" which is a gzipped tarball of native executables. Not sure if wasm-to-oci would work for trying that but I can always make a bigger Wasm, or try the same thing with oras

**Comment by flavio** [comment]
@brooksmtownsend I had a 8Mb Wasm module around (thanks SwiftWasm :wink:) that I successfully uploaded in the past using oci-distribution.

I tried to upload it again, and it failed with the same error message you got.

Long story short... the OCI distribution spec supports [two ways to upload data](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#push=): chunked or monolithic.

The old code of oci-distribution implemented only monolithic push. These have probably a pretty high upload size limit, that's why in the past my 8Mb module was successfully pushed.

The recent code of oci-distribution supports also chunked uploads and perform them by default. The chunked upload code however had a bug, it didn't actually send multiple chunks but it instead sent only one single chunk with the whole data. It was basically a monolithic push in disguise.

As we have seen, GitHub (and probably other registries) have a lower size limit for the individual chunked uploads. That's why we ran into this issue.

I've fixed the issue with https://github.com/krustlet/oci-distribution/pull/37 - feel free to take a look at that (even review) and give it a spin.

I've tested the code with the new patch and it works fine:

```console 
cargo run --example wasm -- push big-wasm.wasm ghcr.io/flavio/big-wasm:latest
```

Pulling the data back and checking the checksums proves the data has been uploaded correctly.

**Comment by brooksmtownsend** [comment]
@flavio with your fix I was able to push a 32MB wasmCloud capability provider in a single "layer"
```
     Running `target/release/wash reg push 'ghcr.io/brooksmtownsend/httpserver:0.5.0' ./httpserver.par.gz -a 'org.opencontainers.image.source=https://github.com/brooksmtownsend/httpserver'`
>>> ⢃⠀  Pushing ./httpserver.par.gz to ghcr.io/brooksmtownsend/httpserver:0.5.0 ...
OciImageManifest { schema_version: 2, media_type: None, config: OciDescriptor { media_type: "application/vnd.wasmcloud.provider.archive.config", digest: "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a", size: 2, urls: None, annotations: None }, layers: [OciDescriptor { media_type: "application/vnd.wasmcloud.provider.archive.layer.v1+par", digest: "sha256:66b76f81b731151491b8913fb67e3f83cf20b020245cf9b2f95426b42cce3e59", size: 31891042, urls: None, annotations: None }], annotati

🚿 Successfully validated and pushed to ghcr.io/brooksmtownsend/httpserver:0.5.0
```

Once your fix is merged in we'll get rid of the custom layering and just use the automatic chunking :)

**Comment by flavio** [comment]
I just tagged the v0.9.1 release. The GH action is going to publish the new crate in a matter of minutes

**Comment by brooksmtownsend** [comment]
@flavio looks like you may have forgotten to bump to `v0.9.1` in `Cargo.toml` 😄

**Comment by thomastaylor312** [comment]
Closed in #37
